README
For now, just one single screensaver for GNOME.




debian/changelog
krigsavers (0.1.1) intrepid; urgency=low

  * Fixed silly bug with the number of the month.

 -- Kristoffer Grönlund <kristoffer.gronlund@purplescout.se>  Mon, 23 Mar 2009 11:53:46 +0100

krigsavers (0.1.0) UNRELEASED; urgency=low

  * Initial release.

 -- Kristoffer Grönlund <kristoffer.gronlund@purplescout.se>  Sat, 14 Mar 2009 17:10:25 +0100


util.h
#ifdef UTIL_H__12234
#error Can only be included once
#endif

#ifndef UTIL_H__12234
#define UTIL_H__12234

#include <cairo.h>
#include <gtk/gtk.h>
#include <stdlib.h>
#include <time.h>
#include <math.h>
#include <string.h>
#include <glib.h>
#include "gs-theme-window.h"

typedef struct rgba_
{
	double r, g, b, a;
} rgba_t;

typedef struct circle_
{
	double x;
	double y;
	double r;
	double dx;
	double dy;
	rgba_t fill;
	rgba_t stroke;
	double t;
} circle_t;

static rgba_t
color(unsigned int rgba)
{
	rgba_t clr;
	clr.a = ((double)((rgba>>24)&0xff))/255.5;
	clr.r = ((double)((rgba>>16)&0xff))/255.5;
	clr.g = ((double)((rgba>> 8)&0xff))/255.5;
	clr.b = ((double)((rgba    )&0xff))/255.5;
	return clr;
}

static GRand* gs_rander = NULL;

static void
free_rander(void)
{
	if (gs_rander)
		g_rand_free(gs_rander);
}

static void
init_random()
{
	if (gs_rander == NULL) {
		gs_rander = g_rand_new_with_seed(time(0));
		atexit(free_rander);
	}
}

static double
next_random()
{
	init_random();
	return g_rand_double_range(gs_rander, -1.0, 1.0);
}

static int
next_random_int_range(int begin, int end)
{
	init_random();
	return g_rand_int_range(gs_rander, begin, end);
}

#endif/*UTIL_H__12234*/


epochsaver.c
#include <cairo.h>
#include <gtk/gtk.h>
#include <stdlib.h>
#include <time.h>
#include <string.h>
#include "gs-theme-window.h"

static gchar *geometry = NULL;
static gboolean countdown = FALSE;
static gboolean g_variant1 = FALSE;
static gboolean g_colour_blue = FALSE;

static GOptionEntry options[] = {
	{"geometry",
	 0,
	 0,
	 G_OPTION_ARG_STRING,
	 &geometry,
	 "The initial size and position of window", "WIDTHxHEIGHT+X+Y"},
	{"countdown",
	 'c',
	 0,
	 G_OPTION_ARG_NONE,
	 &countdown,
	 "If set, the clock counts down to 2000000000", NULL},

	{"variant1",
	 0,
	 0,
	 G_OPTION_ARG_NONE,
	 &g_variant1,
	 "clock variant 1", NULL},

	{"blue",
	 0,
	 0,
	 G_OPTION_ARG_NONE,
	 &g_colour_blue,
	 "blue colour scheme", NULL},
	
       {NULL}
};


static gboolean
on_expose_event(GtkWidget      *widget,
		GdkEventExpose *event,
		gpointer        data)
{
	cairo_t *cr;
	time_t now;
	struct tm *now_tm;
	char snow[256];
	char smax[256];
	char* ptr;
	gint x, y;
	gint width, height;
	int invtim;
	cairo_text_extents_t extents;
	cairo_matrix_t mtx;

	now = time(0);
	now_tm = localtime(&now);

	if (countdown) {
		strftime(snow, 256, "%s", now_tm);
		invtim = 2000000000 - atoi(snow);
		sprintf(snow, "%d", invtim);
	}
	else if (g_variant1) {
		strftime(snow, 256, "%T", now_tm);
	}
	else {
		strftime(snow, 256, "%s", now_tm);
	}

	strcpy(smax, snow);
	for (ptr = smax; *ptr; ++ptr) {
		*ptr = '9';
	}

	cr = gdk_cairo_create(widget->window);

	gdk_window_get_position(widget->window, &x, &y);
	gdk_drawable_get_size(widget->window, &width, &height);

	double bg[3] = {0.0, 0.0, 0.0};
	double fg[4] = {1.0, 0.4, 0.0, 1.0};
	double fg_shadow[4] = {1.0, 0.4, 0.0, 0.2};

	if (g_colour_blue) {
		fg[0] = 0.0;
		fg[2] = 1.0;
		fg_shadow[0] = 0.0;
		fg_shadow[2] = 1.0;
	}

	cairo_set_source_rgb (cr, bg[0], bg[1], bg[2]);
	cairo_paint (cr);

	cairo_set_source_rgba(cr, fg[0], fg[1], fg[2], fg[3]);
	cairo_select_font_face(cr, "Sans", CAIRO_FONT_SLANT_NORMAL,
			       CAIRO_FONT_WEIGHT_NORMAL);
	cairo_set_font_size(cr, 80.0);


	cairo_text_extents(cr,
			   smax,
			   &extents);

	cairo_move_to(cr, (width - extents.width)/2.0, (height - extents.height)/2.0);


	cairo_show_text(cr, snow);

	cairo_set_source_rgba(cr, fg_shadow[0], fg_shadow[1], fg_shadow[2], fg_shadow[3]);
	cairo_move_to(cr, (width - extents.width)/2.0, (height - extents.height)/2.0);
	cairo_matrix_init_scale(&mtx, 80.0, -80.0);
	cairo_set_font_matrix(cr, &mtx);

	cairo_show_text(cr, snow);

	cairo_destroy(cr);

	return FALSE;
}

static gint
timeout_callback (gpointer data)
{
	GtkWidget *darea = (GtkWidget*)data;
	gtk_widget_queue_draw (darea);
	return TRUE;
}

int
main (int argc, char *argv[])
{
	GtkWidget *window;
	GtkWidget *darea;
	GError *error;

	error = NULL;

	gtk_init_with_args (&argc, &argv,
			    "epochsaver",
			    options, NULL, &error);

	window = gs_theme_window_new();
	darea = gtk_drawing_area_new();
	gtk_container_add(GTK_CONTAINER (window), darea);

	g_signal_connect(darea, "expose-event",
			 G_CALLBACK (on_expose_event), NULL);
	g_signal_connect(window, "destroy",
			 G_CALLBACK (gtk_main_quit), NULL);

	g_timeout_add (1000, timeout_callback, darea);

	if ((geometry == NULL)
	    || !gtk_window_parse_geometry (GTK_WINDOW (window), geometry))
		gtk_window_set_default_size (GTK_WINDOW (window), 640, 480);

	gtk_widget_show_all(window);

	gtk_main();

	return 0;
}



swarmsaver.c
#include <cairo.h>
#include <gtk/gtk.h>
#include <stdlib.h>
#include <time.h>
#include <math.h>
#include <string.h>
#include <glib.h>
#include "gs-theme-window.h"

static gchar *geometry = NULL;
static gboolean countdown = FALSE;

static GOptionEntry options[] = {
	{"geometry",
	 0,
	 0,
	 G_OPTION_ARG_STRING,
	 &geometry,
	 "The initial size and position of window", "WIDTHxHEIGHT+X+Y"},
	{"countdown",
	 'c',
	 0,
	 G_OPTION_ARG_NONE,
	 &countdown,
	 "If set, the clock counts down to 1234567890", NULL},

	{NULL}
};

#define MAX_SWARMERS 1000
#define NUM_CIRCLES 4
#define UPDATE_FREQ 10

typedef struct rgba_
{
	double r, g, b, a;
} rgba_t;

typedef struct circle_
{
	double x;
	double y;
	double r;
	double stroke_w;
	rgba_t fill;
	rgba_t stroke;
	int i;
	int on;
} circle_t;

typedef struct swarmer_
{
	double x;
	double y;
	double dx;
	double dy;
	double r;
	double distance;
	double age;
	circle_t* circle;
} swarmer_t;

static rgba_t
color(unsigned int rgba)
{
	rgba_t clr;
	clr.a = ((double)((rgba>>24)&0xff))/255.5;
	clr.r = ((double)((rgba>>16)&0xff))/255.5;
	clr.g = ((double)((rgba>> 8)&0xff))/255.5;
	clr.b = ((double)((rgba    )&0xff))/255.5;
	return clr;
}

GRand* rander = NULL;
static double
next_random()
{
	if (rander == NULL)
		rander = g_rand_new_with_seed(time(0));
	return g_rand_double_range(rander, 0.0, 2.0) - 1.0;
}

static double
next_random_01()
{
	if (rander == NULL)
		rander = g_rand_new_with_seed(time(0));
	return g_rand_double_range(rander, 0, 1.0);
}

#define CIRCLE_RADIUS 0.03
#define CIRCLE_SIDESPACE 0.3
#define CIRCLE_SPACING ((1.0 - ((CIRCLE_SIDESPACE + CIRCLE_RADIUS)*2.0))/(NUM_CIRCLES-1.0))

static void
calc_circle_pos(int i, double* x, double* y)
{
	*x = CIRCLE_SIDESPACE + CIRCLE_RADIUS + (double)i * CIRCLE_SPACING;
	*y = 0.5;
}

static circle_t
gen_circle(int i)
{
	circle_t c;
	calc_circle_pos(i, &c.x, &c.y);
	c.r = CIRCLE_RADIUS;
	c.on = 0;
	c.i = i;
	c.stroke_w = 2.0;
	c.fill = color(0xffffffff);
	c.stroke = color(0xff1c222f);
	return c;
}

static void
gen_swarmer(circle_t* circle);

static double secounter = 0.0;

static void
update_circle(circle_t* c, double dt)
{
	int sec;
	secounter += dt/((double)NUM_CIRCLES * 4.0);
	if ((int)secounter >= NUM_CIRCLES)
		secounter = 0.0;
	sec = (int)secounter;

	int n, i;
	//if (t % (c->i + 2)) {
	//if ((t&0x3) == ((i+1)&0x3)) {
	if (sec == c->i) {
		c->on = 1;
		c->fill = color(0xffffffff);
		c->stroke = color(0xff222222);
        
		n = (int)(next_random()*10.0) - 8;
		for (i = 0; i < n; ++i) {
			gen_swarmer(c);
		}
	} else {
		c->on = 0;
		c->fill = color(0xff000000);
		c->stroke = color(0xff000000);
	}
}

static void
draw_circle(cairo_t* cr, gint width, gint height, circle_t* c)
{
	cairo_set_source_rgba(cr, c->fill.r, c->fill.g, c->fill.b, c->fill.a);
	cairo_arc(cr, c->x*width, c->y*(double)height, c->r*(double)width, 0.0, 2 * M_PI);
	//cairo_stroke_preserve(cr);
	cairo_fill_preserve(cr);
	cairo_set_source_rgba(cr, c->stroke.r, c->stroke.g, c->stroke.b, c->stroke.a);
	cairo_set_line_width(cr, c->stroke_w);
	cairo_stroke(cr);
}

circle_t circles[NUM_CIRCLES];
GPtrArray* swarmers = NULL;

static void
update_swarmer(swarmer_t* s, double dt);

double safesqrt(double st)
{
	if (st < 0.00001)
		return 0.00316;
	return sqrt(st);
}

static swarmer_t* swarmer_freelist = NULL;

static swarmer_t* 
alloc_swarmer()
{
	swarmer_t* ret = NULL;
	if (swarmer_freelist) {
		ret = swarmer_freelist;
		swarmer_freelist = *((swarmer_t**)swarmer_freelist);
	}
	else {
		ret = (swarmer_t*)malloc(sizeof(swarmer_t));
	}
	return ret;
}

static void
free_swarmer(swarmer_t* s)
{
	*((swarmer_t**)s) = swarmer_freelist;
	swarmer_freelist = s;
}

static void
gen_swarmer(circle_t* circle)
{
	if (swarmers->len >= MAX_SWARMERS)
		return;

	double dx, dy;
	double dist = CIRCLE_RADIUS + next_random_01()*0.05;
	swarmer_t* s = alloc_swarmer();
	s->distance = dist;
	s->circle = circle;
	s->x = circle->x;
	s->y = circle->y;
	s->age = 0.0;
	dx = next_random()*(CIRCLE_RADIUS+0.05);
	dy = safesqrt(dist*dist - dx*dx) * (next_random()>0.0 ? -1.0 : 1.0);

	s->dx = dx;
	s->dy = dy;
	s->x += dx;
	s->y += dy;
    
	update_swarmer(s, 0.0);
	g_ptr_array_add(swarmers, s);
}

static double
sqr_distance(double x1, double y1, double x2, double y2)
{
	return ((x1 - x2) * (x1 - x2)) + ((y1 - y2) * (y1 - y2));
}

static double
distance2D(double x1, double y1, double x2, double y2)
{
	double dist2 = sqr_distance(x1, y1, x2, y2);
	return safesqrt(dist2);
}

void scan_vicinity(swarmer_t* s,
		   circle_t** evil,
		   swarmer_t** friend,
		   swarmer_t** enemy, 
		   double* nearest_evil_circle, 
		   double* nearest_friend, 
		   double* nearest_enemy)
{
	double evil_dist = 1000.0;
	double friend_dist = 1000.0;
	double enemy_dist = 1000.0;
	int i;
	circle_t* circ = NULL;
	swarmer_t* s2 = NULL;
    
	*evil = NULL;
	*friend = NULL;
	*enemy = NULL;
    
	for (i = 0; i < NUM_CIRCLES; ++i) {
		circ = circles + i;
		if (circ == s->circle)
			continue;

		double dist = sqr_distance(s->x, circ->x, s->y, circ->y);
		if (evil_dist > dist) {
			evil_dist = dist;
			*evil = circ;
		}
	}
    
	int top = swarmers->len > 15 ? 15 : swarmers->len;
	for (i = 0; i < top; ++i) {
		s2 = g_ptr_array_index(swarmers, i);
		if (s2 == s)
			continue;
		double dist = sqr_distance(s->x, s2->x, s->y, s2->y);
		if (s2->circle == s->circle) {
			if (friend_dist > dist) {
				friend_dist = dist;
				*friend = s2;
			}
		}
		else {
			if (enemy_dist > dist) {
				enemy_dist = dist;
				*enemy = s2;
			}
		}
	}
    
	*nearest_evil_circle = safesqrt(evil_dist);
	*nearest_friend = safesqrt(friend_dist);
	*nearest_enemy = safesqrt(enemy_dist);
}

static void
update_swarmer(swarmer_t* s, double dt)
{
	circle_t* evil;
	swarmer_t* friend;
	swarmer_t* enemy;
	double nearest_evil;
	double nearest_friend;
	double nearest_enemy;
    
	if (dt) {
		scan_vicinity(s, &evil, &friend, &enemy, &nearest_evil, &nearest_friend, &nearest_enemy);
    
		// light attracts * 0.5
		// friend attracts * 0.25
		// evil repels * 0.7
		// enemy repels * 0.9
		double fear_x = 0.0;
		double fear_y = 0.0;
		double lust_x = 0.0;
		double lust_y = 0.0;
    
		double dx, dy;
		dx = s->dx + next_random()*0.05;
		dy = s->dy + next_random()*0.05;
		if (s->circle->on && s->distance < CIRCLE_RADIUS*1.3) {
			dx -= (s->circle->x - s->x) * 0.2;
			dy -= (s->circle->y - s->x) * 0.2;
		}
		else if (s->circle->on && s->distance < 0.3) {
			dx += (s->circle->x - s->x) * 0.2;
			dy += (s->circle->y - s->y) * 0.2;
		}
		if (nearest_friend < 0.1) {
			dx -= (friend->x - s->x) * 0.2;
			dy -= (friend->y - s->y) * 0.2;
		}
		else if (nearest_friend < 0.5) {
			dx += (friend->x - s->x) * 0.2;
			dy += (friend->y - s->y) * 0.2;
		}
		if (evil->on && nearest_evil < 0.2) {
			dx -= (evil->x - s->x) * 0.2;
			dy -= (evil->y - s->y) * 0.2;
		}
		if (s->circle->on) {
			if (nearest_enemy < 0.2) {
				dx += (enemy->x - s->x) * 0.05;
				dy += (enemy->y - s->y) * 0.05;
			}
		} else {
			if (nearest_enemy < 0.8) {
				dx += (enemy->x - s->x) * 0.1;
				dy += (enemy->y - s->y) * 0.1;
			}   
		}
    
		//dx += next_random()*s->age;
		//dy += next_random()*s->age;
		dx *= dt;
		dy *= dt;
		s->age += dt;
		s->dx = dx;
		s->dy = dy;
		s->x += dx;
		s->y += dy;
	}
	s->distance = distance2D(s->x, s->y, s->circle->x, s->circle->y);
}

static void
draw_swarmer(cairo_t* cr, gint width, gint height, swarmer_t* s)
{
	double intensity;
	rgba_t clr;
	switch (s->circle->i)
	{
        case 0: clr = color(0xFFFEE169); break;
        case 1: clr = color(0xFFCDD452); break;
        case 2: clr = color(0xFFF9722E); break;
        default:
        case 3: clr = color(0xFFC9313D); break;
	};
	double agelim = (s->age > 20.0 ? 20.0 : s->age) * 0.1; 
	intensity = 1.0 - (s->distance/0.25) + agelim*0.25;

	cairo_set_source_rgba(cr, clr.r*intensity, clr.g*intensity, clr.b*intensity, clr.a);
	cairo_rectangle(cr, s->x*width - 0.5 - agelim, s->y*(double)height - 0.5 - agelim, 1 + agelim*2.0, 1 + agelim*2.0);
	cairo_arc(cr, s->x*width, s->y*height, 1.0 + agelim*2.0, 0.0, 2 * M_PI);
	cairo_fill(cr);
}

static gboolean
on_expose_event(GtkWidget      *widget,
		GdkEventExpose *event,
		gpointer        data)
{
	cairo_t *cr;
	gint x, y, width, height;
	int i;
	cr = gdk_cairo_create(widget->window);

	gdk_window_get_position(widget->window, &x, &y);
	gdk_drawable_get_size(widget->window, &width, &height);

	
	cairo_set_source_rgb (cr, 0, 0, 0);
	cairo_paint (cr);

	for (i = 0; i < NUM_CIRCLES; ++i) {
		draw_circle(cr, width, height, circles + i);
	}
	
	for (i = 0; i < swarmers->len; ++i) {
		if (g_ptr_array_index(swarmers, i)) {
			draw_swarmer(cr, width, height, (swarmer_t*)g_ptr_array_index(swarmers, i));
		}
	}

	cairo_destroy(cr);

	return FALSE;
}

static void
update_swarmer_cb(gpointer swarmer, gpointer dt)
{
	update_swarmer((swarmer_t*)swarmer, *(double*)dt);
}

static gint
timeout_callback (gpointer data)
{
	swarmer_t* swarmer;
	double dt;
	int i;
	dt = (double)UPDATE_FREQ / 1000.0;
	for (i = 0; i < NUM_CIRCLES; ++i) {
		update_circle(circles + i, dt);
	}
	g_ptr_array_foreach(swarmers, update_swarmer_cb, &dt);
	for (i = 0; i < swarmers->len; ++i) {
		swarmer = (swarmer_t*)g_ptr_array_index(swarmers, i);
		if (swarmer->distance > 0.25) {
			free(swarmer);
			g_ptr_array_remove_index_fast(swarmers, i);
			--i;
		}
	}
	GtkWidget *darea = (GtkWidget*)data;
	gtk_widget_queue_draw (darea);
	return TRUE;
}

void free_swarmer_cb(gpointer swarmer, gpointer user_data)
{
	free_swarmer((swarmer_t*)swarmer);
}

int
main (int argc, char *argv[])
{
	GtkWidget *window;
	GtkWidget *darea;
	GError *error;
	int i;


	error = NULL;

	gtk_init_with_args (&argc, &argv,
			    "circlesaver",
			    options, NULL, &error);

	swarmers = g_ptr_array_sized_new(1000);

	window = gs_theme_window_new();
	darea = gtk_drawing_area_new();
	gtk_container_add(GTK_CONTAINER (window), darea);

	g_signal_connect(darea, "expose-event",
			 G_CALLBACK (on_expose_event), NULL);
	g_signal_connect(window, "destroy",
			 G_CALLBACK (gtk_main_quit), NULL);

	for (i = 0; i < NUM_CIRCLES; ++i) {
		circles[i] = gen_circle(i);
	}
	
	g_timeout_add_full (G_PRIORITY_HIGH, UPDATE_FREQ, timeout_callback, darea, NULL);
	//g_timeout_add (UPDATE_FREQ, timeout_callback, darea);

	if ((geometry == NULL)
	    || !gtk_window_parse_geometry (GTK_WINDOW (window), geometry))
		gtk_window_set_default_size (GTK_WINDOW (window), 640, 480);

	gtk_widget_show_all(window);

	gtk_main();

	g_ptr_array_foreach(swarmers, &free_swarmer_cb, NULL);
	g_ptr_array_free(swarmers, TRUE);
    
	if (rander)
		g_rand_free(rander);

	return 0;
}



crisisaver.c
#include "util.h"
#include <ctype.h>

static gchar *geometry = NULL;
static int selectedtheme = 0;

const char* number2string[] = { "zero", "one", "two", "three", "four", "five", "six", "seven", "eight",
			"nine", "ten", "eleven", "twelve", "thirteen", "fourteen", "fifteen", "sixteen", "seventeen",
			"eighteen", "nineteen", "twenty", "twentyone", "twentytwo", "twentythree",
			"twentyfour", "twentyfive", "twentysix", "twentyseven", "twentyeight", "twentynine",
			"thirty", "thirtyone" };

static GOptionEntry options[] = {
	{"geometry",
	 0,
	 0,
	 G_OPTION_ARG_STRING,
	 &geometry,
	 "The initial size and position of window", "WIDTHxHEIGHT+X+Y"},
 	{"theme",
	 't',
	 0,
	 G_OPTION_ARG_INT,
	 &selectedtheme,
	 "Color theme", NULL},

       {NULL}
};

typedef struct ColorTheme_t
{
	const char* name;
	rgba_t bg;
	rgba_t fg0;
	rgba_t fg1;
	rgba_t fg2;
	rgba_t fg3;
} ColorTheme;

#define COLOR4i(r,g,b,a) { (double)(r)/255.0, (double)(g)/255.0, (double)(b)/255.0, (double)(a)/255.0 }

static ColorTheme g_themes[] = {
	{
		.name = "cinnamon",
		.bg = COLOR4i(32, 30, 24, 255),
		.fg1 = COLOR4i(42, 39, 32, 255),
		.fg0 = COLOR4i(245, 183, 65, 255),
		.fg3 = COLOR4i(68, 64, 57, 255),
		.fg2 = COLOR4i(246, 254, 255, 255),
	},
	{
		.name = "spice",
		.bg = COLOR4i(54, 41, 35, 255),
		.fg0 = COLOR4i(145, 130, 73, 255),
		.fg1 = COLOR4i(255, 245, 178, 255),
		.fg2 = COLOR4i(212, 192, 120, 255),
		.fg3 = COLOR4i(163, 34, 32, 255),
	},
	{
		.name = "sunshine",
		.bg = COLOR4i(112, 120, 100, 255),
		.fg0 = COLOR4i(193, 215, 78, 255),
		.fg1 = COLOR4i(245, 255, 124, 255),
		.fg2 = COLOR4i(223, 230, 180, 255),
		.fg3 = COLOR4i(166, 184, 156, 255),
	},
	{
		.name = "mexican",
		.bg = COLOR4i(36, 19, 35, 255),
		.fg0 = COLOR4i(43, 25, 49, 255),
		.fg1 = COLOR4i(82, 36, 55, 255),
		.fg2 = COLOR4i(191, 158, 165, 255),
		.fg3 = COLOR4i(248, 201, 207, 255),
	},
	{
		.name = "paragon",
		.bg = COLOR4i(37, 35, 38, 255),
		.fg0 = COLOR4i(242, 242, 242, 255),
		.fg1 = COLOR4i(242, 224, 189, 255),
		.fg2 = COLOR4i(242, 95, 41, 255),
		.fg3 = COLOR4i(28, 25, 38, 255),
	},
	{
		.name = "fauxberry",
		.bg = COLOR4i(64, 60, 61, 255),
		.fg0 = COLOR4i(166, 70, 78, 255),
		.fg1 = COLOR4i(140, 130, 116, 255),
		.fg2 = COLOR4i(242, 218, 189, 255),
		.fg3 = COLOR4i(89, 45, 53, 255),
	},
	{
		.name = "leaf",
		.bg = COLOR4i(50, 56, 54, 255),
		.fg0 = COLOR4i(186, 209, 181, 255),
		.fg1 = COLOR4i(219, 232, 207, 255),
		.fg2 = COLOR4i(240, 247, 232, 255),
		.fg3 = COLOR4i(255, 254, 245, 255),
	}


};
#define NUM_THEMES (sizeof(g_themes)/sizeof(g_themes[0]))

static ColorTheme* g_theme = g_themes + 1;

void set_source_rgb(cairo_t* cr, rgba_t* clr)
{
	cairo_set_source_rgb(cr, clr->r, clr->g, clr->b);
}

void set_source_rgba(cairo_t* cr, rgba_t* clr)
{
	cairo_set_source_rgba(cr, clr->r, clr->g, clr->b, clr->a);
}

void set_source_rgba2(cairo_t* cr, rgba_t* clr, double a)
{
	cairo_set_source_rgba(cr, clr->r, clr->g, clr->b, a);
}

void draw_text(cairo_t* cr, const char* text, double x, double y, double size, rgba_t* clr)
{
	cairo_save(cr);
	set_source_rgba(cr, clr);
	cairo_select_font_face(cr, "Sans", CAIRO_FONT_SLANT_NORMAL,
			       CAIRO_FONT_WEIGHT_NORMAL);
	cairo_set_font_size(cr, size);
	cairo_move_to(cr, x, y);
	cairo_show_text(cr, text);
	//cairo_close_path(cr);
	cairo_restore(cr);
}

typedef struct Segment_
{
	double start, end;
} segment_t;

void random_shuffle(int* begin, int* end)
{
	int i, tmp, pick;
	int len = end - begin;
	for (i = len-1; i > 0; --i) {
		pick = g_random_int_range(0, i);
		tmp = begin[i];
		begin[i] = begin[pick];
		begin[pick] = tmp;
	}
}

static segment_t g_segments_sec[60];
static segment_t g_segments_min[60];
static segment_t g_segments_hou[24];
static int g_seeded = 0;

void seed_segments(segment_t* list, int size)
{
	int idx[60];
	int i;
	double offset, len;
	for (i = 0; i < size; ++i) {
		idx[i] = i;
	}
	random_shuffle(idx, idx + size);

	offset = 0.0;
	len = (2.0*M_PI)/(double)size;
	for (i = 0; i < size; ++i) {
		list[idx[i]].start = offset + len*0.05;
		list[idx[i]].end = offset + len*0.9;
		offset = (i+1.0)*len;
	}
	g_seeded = 1;
}

void draw_segment(cairo_t* cr, rgba_t* stroke, rgba_t* fill, double x, double y, double r, double width, segment_t* segment, double alpha)
{
	//double x = width/2 + width/5;
	//double y = height/2;
	//double r = width/4;
	//double stroke_w = 120;
	cairo_save(cr);
	cairo_move_to(cr, x + r * cos(segment->start), y + r * sin(segment->start));
	//set_source_rgba2(cr, stroke, 0.5);
	//cairo_move_to(cr, x - r, y);
	cairo_arc(cr, x, y, r, segment->start, segment->end);

	cairo_move_to(cr, x + r * cos(segment->end*60.0), y + r * sin(segment->end*60.0));
	//cairo_stroke_preserve(cr);
	//cairo_fill_preserve(cr);
	set_source_rgba2(cr, stroke, alpha);
	cairo_set_line_width(cr, width);
	cairo_stroke(cr);
	cairo_restore(cr);
}

void draw_time(cairo_t* cr, int hou, int mins, int secs, double width, double height)
{
	int i;
	if (secs > 59)
		secs = 59; // leap seconds..!
	if (secs == 0 || !g_seeded) {
		seed_segments(g_segments_min, 60);
		seed_segments(g_segments_hou, 24);
		//g_theme = g_themes + g_random_int_range(0, NUM_THEMES);
	}
	seed_segments(g_segments_sec, 60);
	for (i = 0; i < hou; ++i) {
		draw_segment(cr, &g_theme->fg3, 0, width/2 + width/5, height/2, width/4, 120, g_segments_hou + i, 0.5);
	}
	for (i = 0; i < mins; ++i) {
		draw_segment(cr, &g_theme->fg3, 0, width/2 + width/5, height/2, width/4-120, 90, g_segments_min + i, 0.2);
	}
	for (i = 0; i < secs; ++i) {
		draw_segment(cr, &g_theme->fg3, 0, width/2 + width/5, height/2, width/4-240, 60, g_segments_sec + i, 0.1);
	}
}

void str2upper(char* str) {
	while (*str) {
		*str = toupper(*str);
		str++;
	}
}

static gboolean
on_expose_event(GtkWidget      *widget,
		GdkEventExpose *event,
		gpointer        data)
{
	cairo_t *cr;
	time_t now;
	struct tm *now_tm;
	char cur_day[256];
	char cur_date[256];
	char cur_time[256];
	gint x, y;
	gint width, height;
	int invtim;
	cairo_text_extents_t extents;
	cairo_matrix_t mtx;

	now = time(0);
	now_tm = localtime(&now);

	strftime(cur_day, 256, "%A ", now_tm);
	strcat(cur_day, number2string[now_tm->tm_mday]);
	strftime(cur_date, 256, "%B %Y", now_tm);
	strftime(cur_time, 256, "%H:%M", now_tm);

	str2upper(cur_day);
	str2upper(cur_date);
	str2upper(cur_time);

	cr = gdk_cairo_create(widget->window);

	gdk_window_get_position(widget->window, &x, &y);
	gdk_drawable_get_size(widget->window, &width, &height);

	set_source_rgb (cr, &g_theme->bg);
	cairo_paint (cr);

	draw_time(cr, now_tm->tm_hour, now_tm->tm_min, now_tm->tm_sec, width, height);

	draw_text(cr, cur_day, 100, height-220, 40, &g_theme->fg0);
	draw_text(cr, cur_date, 100, height-163, 50, &g_theme->fg2);
	draw_text(cr, cur_time, 100, height-123, 25, &g_theme->fg1);

	cairo_destroy(cr);

	return FALSE;
}

static gint
timeout_callback (gpointer data)
{
	GtkWidget *darea = (GtkWidget*)data;
	gtk_widget_queue_draw (darea);
	return TRUE;
}

int
main (int argc, char *argv[])
{
	GtkWidget *window;
	GtkWidget *darea;
	GError *error;

	error = NULL;

	gtk_init_with_args (&argc, &argv,
			    "crisisaver",
			    options, NULL, &error);

	if (selectedtheme >= NUM_THEMES)
		selectedtheme = 0;
	g_theme = g_themes + selectedtheme;

	window = gs_theme_window_new();
	darea = gtk_drawing_area_new();
	gtk_container_add(GTK_CONTAINER (window), darea);

	g_signal_connect(darea, "expose-event",
			 G_CALLBACK (on_expose_event), NULL);
	g_signal_connect(window, "destroy",
			 G_CALLBACK (gtk_main_quit), NULL);

	g_timeout_add (1000, timeout_callback, darea);

	if ((geometry == NULL)
	    || !gtk_window_parse_geometry (GTK_WINDOW (window), geometry))
		gtk_window_set_default_size (GTK_WINDOW (window), 640, 480);

	gtk_widget_show_all(window);

	gtk_main();

	return 0;
}

